File: sdk/go/network_policy.go
package axernsdk

import (
	"fmt"
	"net/netip"
	"sort"
	"strings"

	commonv1 "github.com/cofy-x/axern/sdk/go/gen/axern/control/common/v1"
	"golang.org/x/net/idna"
	"google.golang.org/protobuf/proto"
)

const maxNetworkPolicyRules = 256

var forbiddenNetworkPolicyCIDRs = []netip.Prefix{
	netip.MustParsePrefix("0.0.0.0/32"),
	netip.MustParsePrefix("100.100.100.200/32"),
	netip.MustParsePrefix("127.0.0.0/8"),
	netip.MustParsePrefix("169.254.0.0/16"),
	netip.MustParsePrefix("192.0.0.192/32"),
	netip.MustParsePrefix("224.0.0.0/4"),
	netip.MustParsePrefix("::/128"),
	netip.MustParsePrefix("::1/128"),
	netip.MustParsePrefix("fe80::/10"),
	netip.MustParsePrefix("ff00::/8"),
}

type EgressProtocol string

const (
	EgressProtocolTCP EgressProtocol = "tcp"
	EgressProtocolUDP EgressProtocol = "udp"
)

type PortRange struct {
	Start uint32
	End   uint32
}

type CIDRRule struct {
	CIDR     string
	Protocol EgressProtocol
	Ports    []PortRange
}

type NetworkPolicy struct {
	wire *commonv1.NetworkEgressPolicy
}

func NewStrictNetworkPolicy(domains []string, cidrRules []CIDRRule) (*NetworkPolicy, error) {
	normalizedDomains, err := normalizePolicyDomains(domains)
	if err != nil {
		return nil, err
	}
	normalizedCIDRs, err := normalizePolicyCIDRs(cidrRules)
	if err != nil {
		return nil, err
	}
	if len(normalizedDomains)+len(normalizedCIDRs) > maxNetworkPolicyRules {
		return nil, fmt.Errorf("network policy may contain at most %d rules", maxNetworkPolicyRules)
	}
	return &NetworkPolicy{wire: &commonv1.NetworkEgressPolicy{Policy: &commonv1.NetworkEgressPolicy_Strict{Strict: &commonv1.StrictEgressPolicy{AllowedDomains: normalizedDomains, AllowedCidrs: normalizedCIDRs}}}}, nil
}

func AllowDomainNetworkPolicy(domains ...string) (*NetworkPolicy, error) {
	return NewStrictNetworkPolicy(domains, nil)
}

func DenyDNSNetworkPolicy(domains ...string) (*NetworkPolicy, error) {
	normalized, err := normalizePolicyDomains(domains)
	if err != nil {
		return nil, err
	}
	if len(normalized) == 0 {
		return nil, fmt.Errorf("deny DNS policy requires at least one domain")
	}
	return &NetworkPolicy{wire: &commonv1.NetworkEgressPolicy{Policy: &commonv1.NetworkEgressPolicy_DnsDeny{DnsDeny: &commonv1.DnsDenyPolicy{DeniedDomains: normalized}}}}, nil
}

func DenyAllNetworkPolicy() *NetworkPolicy {
	policy, _ := NewStrictNetworkPolicy(nil, nil)
	return policy
}

func (p *NetworkPolicy) proto() *commonv1.NetworkEgressPolicy {
	if p == nil || p.wire == nil {
		return nil
	}
	return proto.Clone(p.wire).(*commonv1.NetworkEgressPolicy)
}

func normalizePolicyDomains(values []string) ([]string, error) {
	seen := map[string]struct{}{}
	result := make([]string, 0, len(values))
	for _, value := range values {
		raw := strings.ToLower(strings.TrimSpace(value))
		wildcard := strings.HasPrefix(raw, "*.")
		if wildcard {
			raw = strings.TrimPrefix(raw, "*.")
		}
		raw = strings.TrimSuffix(raw, ".")
		if raw == "" || strings.ContainsAny(raw, "/:@?#*") {
			return nil, fmt.Errorf("invalid domain rule %q", value)
		}
		if _, err := netip.ParseAddr(raw); err == nil {
			return nil, fmt.Errorf("domain rule must not be an IP literal: %q", value)
		}
		ascii, err := idna.Lookup.ToASCII(raw)
		if err != nil || len(ascii) > 253 {
			return nil, fmt.Errorf("invalid domain rule %q", value)
		}
		for _, label := range strings.Split(ascii, ".") {
			if label == "" || len(label) > 63 {
				return nil, fmt.Errorf("invalid domain rule %q", value)
			}
		}
		if wildcard {
			ascii = "*." + ascii
		}
		if _, ok := seen[ascii]; !ok {
			seen[ascii] = struct{}{}
			result = append(result, ascii)
		}
	}
	return result, nil
}

func normalizePolicyCIDRs(values []CIDRRule) ([]*commonv1.CIDREgressRule, error) {
	seen := map[string]*commonv1.CIDREgressRule{}
	for index, value := range values {
		prefix, err := netip.ParsePrefix(strings.TrimSpace(value.CIDR))
		if err != nil {
			return nil, fmt.Errorf("CIDR rule %d: %w", index, err)
		}
		prefix = prefix.Masked()
		for _, forbidden := range forbiddenNetworkPolicyCIDRs {
			if prefix.Addr().BitLen() == forbidden.Addr().BitLen() && prefix.Bits() >= forbidden.Bits() && forbidden.Contains(prefix.Addr()) {
				return nil, fmt.Errorf("CIDR rule %d targets a protected address range", index)
			}
		}
		protocol := commonv1.EgressProtocol_EGRESS_PROTOCOL_TCP
		if value.Protocol == EgressProtocolUDP {
			protocol = commonv1.EgressProtocol_EGRESS_PROTOCOL_UDP
		} else if value.Protocol != EgressProtocolTCP {
			return nil, fmt.Errorf("CIDR rule %d protocol must be tcp or udp", index)
		}
		if len(value.Ports) == 0 {
			return nil, fmt.Errorf("CIDR rule %d requires at least one port range", index)
		}
		ports := make([]*commonv1.PortRange, 0, len(value.Ports))
		for _, value := range value.Ports {
			end := value.End
			if end == 0 {
				end = value.Start
			}
			if value.Start == 0 || end < value.Start || end > 65535 {
				return nil, fmt.Errorf("CIDR rule %d ports must be within 1..65535", index)
			}
			ports = append(ports, &commonv1.PortRange{Start: value.Start, End: end})
		}
		sort.Slice(ports, func(i, j int) bool {
			return ports[i].GetStart() < ports[j].GetStart() || ports[i].GetStart() == ports[j].GetStart() && ports[i].GetEnd() < ports[j].GetEnd()
		})
		var keyBuilder strings.Builder
		fmt.Fprintf(&keyBuilder, "%s|%d", prefix, protocol)
		for _, port := range ports {
			fmt.Fprintf(&keyBuilder, "|%d-%d", port.GetStart(), port.GetEnd())
		}
		key := keyBuilder.String()
		seen[key] = &commonv1.CIDREgressRule{Cidr: prefix.String(), Protocol: protocol, Ports: ports}
	}
	keys := make([]string, 0, len(seen))
	for key := range seen {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	result := make([]*commonv1.CIDREgressRule, 0, len(keys))
	for _, key := range keys {
		result = append(result, seen[key])
	}
	return result, nil
}

File: sdk/typescript/src/network-policy.ts
/**
 * @license
 * Copyright 2026 cofy-x
 * SPDX-License-Identifier: Apache-2.0
 */

import { isIP } from "node:net";
import { domainToASCII } from "node:url";

export interface PortRange {
  readonly start: number;
  readonly end: number;
}

export interface CIDRRule {
  readonly cidr: string;
  readonly protocol: "tcp" | "udp";
  readonly ports: readonly PortRange[];
}

export interface StrictNetworkPolicyOptions {
  readonly domains?: readonly string[];
  readonly cidrRules?: readonly CIDRRule[];
}

type NetworkPolicyWire =
  | { strict: { allowed_domains: string[]; allowed_cidrs: Record<string, unknown>[] } }
  | { dns_deny: { denied_domains: string[] } };

const maxRules = 256;

export class NetworkPolicy {
  private constructor(private readonly wire: NetworkPolicyWire) {}

  static strict(options: StrictNetworkPolicyOptions = {}): NetworkPolicy {
    const domains = normalizeDomains(options.domains ?? []);
    const cidrRules = normalizeCIDRRules(options.cidrRules ?? []);
    if (domains.length + cidrRules.length > maxRules) {
      throw new RangeError(`network policy may contain at most ${maxRules} rules`);
    }
    return new NetworkPolicy({
      strict: {
        allowed_domains: domains,
        allowed_cidrs: cidrRules.map((rule) => ({
          cidr: rule.cidr,
          protocol: rule.protocol === "tcp" ? 1 : 2,
          ports: rule.ports.map((port) => ({ start: port.start, end: port.end })),
        })),
      },
    });
  }

  static allowDomains(...domains: string[]): NetworkPolicy {
    return NetworkPolicy.strict({ domains });
  }

  static denyDns(...domains: string[]): NetworkPolicy {
    const normalized = normalizeDomains(domains);
    if (normalized.length === 0) throw new RangeError("denyDns requires at least one domain");
    return new NetworkPolicy({ dns_deny: { denied_domains: normalized } });
  }

  static denyAll(): NetworkPolicy {
    return NetworkPolicy.strict();
  }

  /** @internal */
  toWire(): NetworkPolicyWire {
    return structuredClone(this.wire);
  }
}

export function portRange(start: number, end: number = start): PortRange {
  if (!Number.isInteger(start) || !Number.isInteger(end) || start < 1 || end < start || end > 65535) {
    throw new RangeError("ports must be an inclusive range within 1..65535");
  }
  return Object.freeze({ start, end });
}

export function cidrRule(cidr: string, protocol: "tcp" | "udp", ...ports: PortRange[]): CIDRRule {
  return normalizeCIDRRule({ cidr, protocol, ports });
}

function normalizeDomains(values: readonly string[]): string[] {
  const result: string[] = [];
  const seen = new Set<string>();
  for (const value of values) {
    let raw = value.trim().toLowerCase();
    const wildcard = raw.startsWith("*.");
    if (wildcard) raw = raw.slice(2);
    raw = raw.replace(/\.+$/, "");
    if (raw === "" || /[/:@?#*]/u.test(raw) || isIP(raw) !== 0) {
      throw new TypeError(`invalid domain rule ${JSON.stringify(value)}`);
    }
    const ascii = domainToASCII(raw).toLowerCase();
    const labels = ascii.split(".");
    if (ascii === "" || Buffer.byteLength(ascii, "ascii") > 253 || labels.some((label) => label === "" || label.length > 63)) {
      throw new TypeError(`invalid domain rule ${JSON.stringify(value)}`);
    }
    const normalized = wildcard ? `*.${ascii}` : ascii;
    if (!seen.has(normalized)) {
      seen.add(normalized);
      result.push(normalized);
    }
  }
  return result;
}

function normalizeCIDRRules(values: readonly CIDRRule[]): CIDRRule[] {
  const result: CIDRRule[] = [];
  const seen = new Set<string>();
  for (const value of values) {
    const rule = normalizeCIDRRule(value);
    const key = JSON.stringify(rule);
    if (!seen.has(key)) {
      seen.add(key);
      result.push(rule);
    }
  }
  return result;
}

function normalizeCIDRRule(value: CIDRRule): CIDRRule {
  const cidr = value.cidr.trim();
  const separator = cidr.lastIndexOf("/");
  const address = separator < 0 ? "" : cidr.slice(0, separator);
  const prefix = separator < 0 ? -1 : Number(cidr.slice(separator + 1));
  const family = isIP(address);
  if ((family !== 4 && family !== 6) || !Number.isInteger(prefix) || prefix < 0 || prefix > (family === 4 ? 32 : 128)) {
    throw new TypeError(`invalid CIDR ${JSON.stringify(value.cidr)}`);
  }
  if (isProtectedCIDR(address, prefix, family)) {
    throw new TypeError(`CIDR ${JSON.stringify(value.cidr)} targets a protected address range`);
  }
  if (value.protocol !== "tcp" && value.protocol !== "udp") throw new TypeError("CIDR protocol must be tcp or udp");
  if (value.ports.length === 0) throw new RangeError("CIDR rule requires at least one port range");
  const ports = value.ports.map((port) => portRange(port.start, port.end));
  return Object.freeze({ cidr, protocol: value.protocol, ports: Object.freeze(ports) });
}

function isProtectedCIDR(address: string, prefix: number, family: number): boolean {
  if (family === 4) {
    const value = address.split(".").reduce((result, octet) => (result << 8n) | BigInt(octet), 0n);
    const protectedRanges: readonly [bigint, number][] = [
      [0n, 32],
      [1684301000n, 32], // 100.100.100.200
      [2130706432n, 8], // 127.0.0.0
      [2851995648n, 16], // 169.254.0.0
      [3221225664n, 32], // 192.0.0.192
      [3758096384n, 4], // 224.0.0.0
    ];
    return protectedRanges.some(([base, bits]) => prefix >= bits && (value >> BigInt(32 - bits)) === (base >> BigInt(32 - bits)));
  }
  const canonical = new URL(`http://[${address}]/`).hostname.slice(1, -1).toLowerCase();
  if (prefix >= 128 && (canonical === "::" || canonical === "::1")) return true;
  const first = Number.parseInt(canonical.split(":", 1)[0] || "0", 16);
  return (prefix >= 10 && first >= 0xfe80 && first <= 0xfebf) || (prefix >= 8 && first >= 0xff00 && first <= 0xffff);
}
